uxarray/grid/coordinates.py

```diff
@@ -227,9 +228,86 @@ def _populate_face_centroids(grid, repopulate=False):
         )
 
 
+def _populate_face_centerpoints(grid, repopulate=False):
+    """Calculates the face centerpoints using Welzl's algorithm. It is a
+    randomized algorithm for finding the center and radius of the smallest
+    circle that encloses a set of points. It is here adapted to work on a unit
+    sphere. Also, this algorithm cannot be guaranteed to work on concave
+    polygons.
+
+    Parameters
+    ----------
+    grid : Grid
+        The grid containing the nodes and faces.
+    repopulate : bool, optional
+        Bool used to turn on/off repopulating the face coordinates of the centerpoints, default is False.
+    """
+    # warnings.warn("This cannot be guaranteed to work correctly on concave polygons")
+
+    node_lon = grid.node_lon.values
+    node_lat = grid.node_lat.values
+
+    centerpoint_lat = []
+    centerpoint_lon = []
+
+    face_nodes = grid.face_node_connectivity.values
+    n_nodes_per_face = grid.n_nodes_per_face.values
+
+    # Check if the centerpoints are already populated
+    if "face_lon_ctrpt" not in grid._ds or repopulate:
+        # Construct the centerpoints if there are none stored
+        if "face_x_ctrpt" not in grid._ds:
+            centerpoint_lon, centerpoint_lat = _construct_face_centerpoints(
+                node_lon, node_lat, face_nodes, n_nodes_per_face
+            )
+    # get the cartesian coordinates of the centerpoints
+    ctrpt_x, ctrpt_y, ctrpt_z = _lonlat_rad_to_xyz(centerpoint_lon, centerpoint_lat)
+
+    # set the grid variables for centerpoints
+    if "face_lon_ctrpt" not in grid._ds or repopulate:
+        grid._ds["face_lon_ctrpt"] = xr.DataArray(
+            centerpoint_lon, dims=[ugrid.FACE_DIM], attrs=ugrid.FACE_LON_ATTRS
+        )
+        grid._ds["face_lat_ctrpt"] = xr.DataArray(
+            centerpoint_lat, dims=[ugrid.FACE_DIM], attrs=ugrid.FACE_LAT_ATTRS
+        )
+
+    if "face_x_ctrpt" not in grid._ds or repopulate:
+        grid._ds["face_x_ctrpt"] = xr.DataArray(
+            ctrpt_x, dims=[ugrid.FACE_DIM], attrs=ugrid.FACE_X_ATTRS
+        )
+
+        grid._ds["face_y_ctrpt"] = xr.DataArray(
+            ctrpt_y, dims=[ugrid.FACE_DIM], attrs=ugrid.FACE_Y_ATTRS
+        )
+
+        grid._ds["face_z_ctrpt"] = xr.DataArray(
+            ctrpt_z, dims=[ugrid.FACE_DIM], attrs=ugrid.FACE_Z_ATTRS
+        )
+
+
 def _construct_face_centroids(node_x, node_y, node_z, face_nodes, n_nodes_per_face):
     """Constructs the xyz centroid coordinate for each face using Cartesian
-    Averaging."""
+    Averaging.
```

Comment:
Can this function be vectorized without the for loop?

Reply:
This should be able to be vectorized in a way similar to the following codes
```python
    # Mask to ignore invalid indices
    mask = np.arange(face_nodes.shape[1])[None, :] < n_nodes_per_face[:, None]

    # Calculate centroids
    centroid_x = np.sum(node_x[face_nodes] * mask, axis=1) / n_nodes_per_face
    centroid_y = np.sum(node_y[face_nodes] * mask, axis=1) / n_nodes_per_face
    centroid_z = np.sum(node_z[face_nodes] * mask, axis=1) / n_nodes_per_face

    return _normalize_xyz(centroid_x, centroid_y, centroid_z)
 ```

Reply:
> This should be able to be vectorized in a way similar to the following codes
> 
> ```python
>     # Mask to ignore invalid indices
>     mask = np.arange(face_nodes.shape[1])[None, :] < n_nodes_per_face[:, None]
> 
>     # Calculate centroids
>     centroid_x = np.sum(node_x[face_nodes] * mask, axis=1) / n_nodes_per_face
>     centroid_y = np.sum(node_y[face_nodes] * mask, axis=1) / n_nodes_per_face
>     centroid_z = np.sum(node_z[face_nodes] * mask, axis=1) / n_nodes_per_face
> 
>     return _normalize_xyz(centroid_x, centroid_y, centroid_z)
> ```
I don't see the code snipped you provided in our repo, where is it from?
This PR is not about face_centroids, in anycase this change can be made if it is not already a part of some other PR.
For this PR, `_construct_face_centerpoints` is hard to vectorize given the call to `smallest_enclosing_circle` because it operates on individual points and not entire set. We can handle it in another PR as it might involve using the `concurrent` package. Another solution is doing this smallest_encolsing_circle for the entire grid at once and it surely complicates the algorithm.